### PR TITLE
Allow dropping cancel_tx when awaiting on Agent::dial and Agent::accept

### PR DESCRIPTION
**Pull request by pheki**
Hey, thanks a lot for this library!

I was connecting on another task with a sent cancel_rx, and dropping the cancel_tx would cancel the operation instead of continuing, as `cancel_rx.recv()` returns None when the channel has been closed. This could also be hit by users ignoring the binding by using _ when binding the Sender, such as:

```rust
// This does not bind the Sender, dropping it immediatly, which would be considered a cancel in the old code.
let (_, cancel_rx) = mpsc::channel(1);
```

A simple way to check the old vs new behavior in ping_pong.rs is to add a drop(cancel_tx) before dialing, such as:

```rust
let (cancel_tx, cancel_rx) = mpsc::channel(1);
drop(cancel_tx);
// Start the ICE Agent. One side must be controlled, and the other must be controlling
let conn: Arc<dyn Conn + Send + Sync> = if is_controlling {
    ice_agent.dial(cancel_rx, remote_ufrag, remote_pwd).await?
} else {
    ice_agent
        .accept(cancel_rx, remote_ufrag, remote_pwd)
        .await?
};
```

Maybe a test could be added checking this behavior, I can give it a try if you want.

### CODE REVIEW

**Comment by rainliu** [comment]
I don’t see any issue in current canceling handling in repo. your chances will make dial and accept ignore drop(cancel_tx)

File: src/agent/agent_transport.rs [hunk 1]
```diff
@@ -27,7 +27,7 @@ impl Agent {
             // block until pair selected
             tokio::select! {
                 _ = on_connected_rx.recv() => {},
-                _ = cancel_rx.recv() => {
+                Some(()) = cancel_rx.recv() => {
```
Comment (rainliu) [inline src/agent/agent_transport.rs:30]: I don't see any justification for this change.

With this change, drop(cancel_tx) won't cancel the dail and accept, which is not acceptable.

**Comment by pheki** [comment]
I think that by dropping the sender you're specifically not sending the cancel message (if I don't want to cancel, why do I need to keep it?), as you can always send `cancel_tx.send(())` (which looks like the supposed way to cancel) before dropping. I've fixed the test to cancel by using the channel instead of dropping it, in case it is approved, sorry about that.

If you keep the current behavior, maybe it should be documented (in dial and accept definitions) that the signal is cancelled not only by sending a message but also by dropping all senders, as I feel like more people may be confused.

**Comment by rainliu** [comment]
even with current handling, you can still cancel it by sending cancel_tx.send(()).

**Comment by metaclips** [comment]
Can we wrap the receiver channel in a `Option` so that users can ignore the cancel request.

```rust
 pub async fn dial(
        &self,
        cancel_rx: Option<mpsc::Receiver<()>>,
        remote_ufrag: String,
        remote_pwd: String,
    ) -> Result<Arc<impl Conn>> {

 if let Some(mut on_connected_rx) = on_connected_rx {
            // block until pair selected
            let is_cancelable = cancel_rx.is_some();

            tokio::select! {
                _ = on_connected_rx.recv() => {},

                _ = async {
                    match cancel_rx{
                        Some(mut e) => e.recv().await,
                        _ => unreachable!(),
                    }
                }, if is_cancelable => {
                    return Err(Error::ErrCanceledByCaller.into());
                }
            }
        }
        Ok(agent_conn)
}
```

So we can call the `dial` method as
```rust
ice_agent.dial(None, remote_ufrag, remote_pwd).await?
```
without creating a cancel context.

If a cancel receiver is dropped, we should close the event.

**Comment by pheki** [comment]
@rainliu Yes, but it is also easy to accidentally cancel, for example:

```
let (_, cancel_rx) = mpsc::channel(1);
// This will be cancelled as the sender has been dropped 
let conn = ice_agent.dial(Some(cancel_rx), remote_ufrag, remote_pwd).await?;
```

```
let (cancel_tx, cancel_rx) = mpsc::channel(1);
tokio::spawn(async {
    // This will be cancelled as the sender has been dropped (it was in another task which doesn't actually want to cancel)
    let conn = ice_agent.dial(Some(cancel_rx), remote_ufrag, remote_pwd).await?;
});
```

```
let (cancel_tx, cancel_rx) = mpsc::channel(1);
let cancel_tx2 = cancel_tx.clone();
tokio::spawn(async {
    async_fn_here().await;
    // This will not actually cancel as there is another sender
    drop(cancel_tx2);
});

let conn = ice_agent.dial(Some(cancel_rx), remote_ufrag, remote_pwd).await?;
```

But I find @metaclips actually pretty cool, since it makes the functions more ergonomic for the general use-case, also tested locally and it passes all tests.

**Comment by metaclips** [comment]
@rainliu apparently this change will be a breaking change anyway we can get it accepted?

**Comment by rainliu** [comment]
I am still not convinced by this change, even adding Option<>,

First, if you accidentally use _ for cancel_tx, you will receive the Error::ErrCanceledByCaller immediately.
```
let (_, cancel_rx) = mpsc::channel(1);
let conn = ice_agent.dial(cancel_rx, remote_ufrag, remote_pwd).await?;
```

Second, if you don't have need to cancel dial/accept, you can just use _cancel_tx, instead of _.
```
let (_cancel_tx, cancel_rx) = mpsc::channel(1);
```

In practical use cases, cancelable is always required in order to make sure there is no deadloop or resource leak.

**Comment by pheki** [comment]
I guess you're right, it needs to be cancelable in production programs.

On the other hand, the cancellation procedure could be implemented by the user (caller) instead of the library, the user could call `tokio::select!` themself (with any condition it decides to implement), that way, cancel_rx wouldn't even be necessary.

The current way is a little more opinionated (which isn't inherently better or worse), favoring a specific cancellation strategy.

**Comment by pheki** [comment]
So, should I close this PR and leave it as it is (less breaking changes)? 

In this case, would you agree with me on opening another PR just document the `cancel_rx` behavior on both methods? Something along the lines of "The operation will be cancelled if `cancel_rx` either receives a message or its channel closes".

**Pull request closed without merging**